### PR TITLE
fix: add Locale.ROOT to bare toLowerCase/toUpperCase calls (release/2026.08)

### PR DESCRIPTION
**Pull request by yingbull**
## Description

Supported-release port of #2425 onto `release/2026.08`. Replaces every bare `toLowerCase()` / `toUpperCase()` in the surveyed files with an explicit `Locale.ROOT` (or a locale-independent equivalent), and hardens the call sites that turned out to be doing real validation. With one deliberate exception noted below, these sites compare ASCII protocol tokens, MIME types, JDBC type names, or internal identifiers — not user-visible display text — so `Locale.ROOT` is the correct and explicit choice for them.

**This branch carries only the fix.** It was cut from `release/2026.08`, not from `develop`, and brings no `develop` history with it. #2425's own branch had `develop` merged into it and cannot be retargeted here without dragging every `develop` commit along.

| File | What's being cased | How it's handled now |
|---|---|---|
| `AddEditDocument2Action` | `.pdf` extension detection (2×) | null-guarded `regionMatches(true, …)` |
| `FrmRecordHelp` | JDBC column type `TINYINT` / `bit` check (3×) | one `isBooleanColumnType` helper over named constants |
| `UtilMisc` | **patient-name display casing** | **deliberately left default-locale** (comment + JavaDoc only) — see below |
| `DemographicContactCreator` | last-name sort comparator | null-safe `toUpperCase(Locale.ROOT).compareTo(…)` — full case mapping preserved |
| `CarlosEncodeTag` | case-insensitive encoding-context dispatch (XSS prevention tag) | `toLowerCase(Locale.ROOT)` |
| `Utility` | `"TODAY"` date-string sentinel (3×) | `TODAY_KEYWORD.equalsIgnoreCase(…)` (null-safe; guarded by `IsEmpty` upstream) |
| `OAuth1ParamParser` | content-type MIME match | `regionMatches` against a named `FORM_URLENCODED` constant |
| `OAuth1SignatureVerifierImplementation` | scheme/host lowercasing, method + percent-encoding uppercasing per RFC 5849 (5×) | `Locale.ROOT`, plus the same named constant |
| `ProviderService` | JSON `"true"` / `"false"` coercion | `toLowerCase(Locale.ROOT)` + exact `equals` (deliberately ASCII-strict) |

### Where this differs from #2425's `develop` branch

- **`CarlosEncodeTag`, not `SafeEncode`.** On this line the encoding-context `switch` still lives in the tag's private `encode()`; `develop` later moved it into `SafeEncode.forContext()`. The fix is applied where the code actually is, and `SafeEncode` is untouched. This site is a genuine bug rather than hygiene: under a Turkish locale `"javaScript".toLowerCase()` yields a dotless ı, the context lookup misses, and the tag throws `JspException` instead of encoding.
- **Regression test added for `active` coercion.** Nothing on either branch tested `search()`'s `active` coercion, and the failure mode this guards is *silent acceptance*: an intermediate revision of #2425 used `equalsIgnoreCase`, and `"falſe".equalsIgnoreCase("false")` is `true`, so a lookalike coerced to `active=false` and returned inactive providers instead of the documented 400. `ProviderServiceRegressionTest` now pins both halves — the lookalike is rejected with 400, and an ASCII case variant (`" FALSE "`) is still accepted and passed through as `false`.
- **Collation regression fixed and pinned (Codex review, 1bac8c2).** #2425's branch had switched `byLastName` to `compareToIgnoreCase`, which changed a user-visible sort order: `String.toUpperCase` maps `ß` → `SS`, but per-`char` folding leaves `ß` as `ß`, so `"Weiß"` moved from sorting as `"WEISS"` to sorting after every Latin letter (verified: `"ßa".toUpperCase(ROOT).compareTo("TA")` is `-1`, `"ßa".compareToIgnoreCase("Ta")` is `107`). Restored `toUpperCase(Locale.ROOT).compareTo(…)`, which is locale-independent, keeps the new null handling, and preserves the previous ordering exactly. `DemographicContactCreatorUnitTest` pins it (sharp s folds to SS, case ignored, nulls first).
- **`UtilMisc.toUpperLowerCase` left default-locale (Codex review, 85ee8cf).** #2425 applied `Locale.ROOT` here too, but every caller formats patient names for display (`appointmentprovideradminday.jsp`, `appointmentgrouprecords.jsp`) — this is the one surveyed site that *is* display text. Under a Turkish default locale the original code renders `IŞIK` as `Işık` (dotless ı, correct) while `Locale.ROOT` renders `Işik` (wrong); on a Canadian locale the two are identical. Reverted to the original call — a **comment-only change to this file** — so the release line has zero behaviour change here; an explicit display locale would be a design change outside a release fix. Both of these Codex findings also apply to #2425's `develop` branch and reach `develop` corrected via this PR's forward-merge.
- **Contract JavaDoc (CodeRabbit review, 30c196f).** Comment-only commit adding JavaDoc to `UtilMisc.toUpperLowerCase` and `Utility.GetSysDate` / `GetSysDateMin` / `GetSysDateMax` (null/blank sentinels, the case-insensitive `"TODAY"` token, parsed shape, and the differing failure modes). It also corrects the inline comment from 85ee8cf, which wrongly claimed an `IMPROPER_UNICODE` suppression covers `toUpperLowerCase`; none exists or is needed, because FindSecBugs only reports a case fold whose result feeds an equality comparison.
- **`TODAY_KEYWORD` constant + `UtilityUnitTest` (SonarCloud `java:S1192`, cfea87d).** The `"TODAY"` literal in `Utility` is now a single named constant, mirroring the `FrmRecordHelp` treatment; the three call sites are otherwise unchanged. `UtilityUnitTest` (7 tests) pins the documented contract: any-case `"TODAY"`, blank-input sentinels, `d/m/yyyy` parsing, `null` vs thrown `Invalid Date` on garbage.

### Carried from the #2425 review cycle

- **`FrmRecordHelp` — `isBooleanColumnType`.** The triplicated `"TINYINT"`/`"bit"` check (SonarCloud `java:S1192`) is one private helper over named constants. It reads `getColumnTypeName(i)` once per column and gives the JDBC-type case fold a single `@SuppressFBWarnings` site.
- **OAuth form-urlencoded check.** The hand-counted length `33` is replaced by a named `FORM_URLENCODED` constant and its own `.length()`. The literal is in fact 33 characters (a review had claimed 32), so behaviour is unchanged.
- **`DemographicContactCreator.byLastName` suppression placement.** The `IMPROPER_UNICODE` suppression moves from the field onto the anonymous class's `compare()`, where FindSecBugs actually reports it; a field-level annotation never covered it.

### Silent-failure review

Each null guard was traced rather than assumed:
- `addDocument`: `fileName2` comes from a validated upload name and is passed to `writeValidatedUpload(...)` *before* the `.pdf` check, so it cannot be null at that point — the guard is dead-but-harmless and behaviour is identical on every reachable path.
- `editDocument`: `fileName` is initialised to `""`; `"".regionMatches(true, -4, ".pdf", 0, 4)` is `false`, matching `"".endsWith(".pdf")`.
- `isBooleanColumnType(null)` returns `false` (was NPE); MySQL/MariaDB never return a null type name, so this is defensive only.
- `ProviderService`: invalid values still throw 400 — nothing is swallowed. Unknown encoding contexts still throw.

**Pre-existing, out of scope, flagged for tickets:**
- `ProviderManager2.search(...)` takes a primitive `boolean active` while `ProviderService.search()` passes a `Boolean` that is `null` when the request omits `active` — an auto-unboxing NPE → 500 on every `/providers/search` without `active`, on both this line and `develop`. Not introduced here and deliberately not widened into a release-line locale fix.
- `Utility.GetSysDateMin` / `GetSysDateMax` parse with `substring(3, 2)` (begin > end), so every non-blank, non-`"TODAY"` input throws. Neither method has an in-tree caller, so nothing reachable is affected. Noted inline in 30c196f and left unchanged.
- `UtilMisc.toUpperLowerCase` duplicates `io.github.carlos_emr.Misc.toUpperLowerCase` (SonarCloud counts the comment-only edits inside it as duplicated lines). Legacy twin classes; not deduplicated here.

## Related Issues

Related to #2425 (same fix, targeted at `develop`; superseded by the forward-merge once this lands).

## Target Branch

Supported fix: topic branch from `release/2026.08` → `release/2026.08`, per `docs/release-process.md` "Supported-release fix cycle". Maintainers forward-merge into `develop`; no duplicate cherry-pick PR is intended.

## How Was This Tested?

Run against this branch, i.e. the release line:

- `mvn compile` — BUILD SUCCESS
- Targeted tests (all classes covering the touched files that exist on this line) — **143 tests, 0 failures, 0 errors** on `c7fd5d4`, including `ProviderServiceRegressionTest` at 5 (3 existing + 2 new)
- Full unit suite (`mvn test -Dgroups=unit`, UTF-8 locale) — **9,344 tests, 0 failures, 0 errors, 3 skipped, BUILD SUCCESS** on `c7fd5d4` (the release line carries fewer tests than `develop`, where the identical diff passed 10,280 / 0 / 0 in #2425)
- `1bac8c2` (collation fix: one production line plus one new test class) — re-ran the two affected classes, `DemographicContactCreatorUnitTest` (3) and `ProviderServiceRegressionTest` (5): **8 / 0 / 0, BUILD SUCCESS**; CI ran the full suite on this head and passed (39/39 checks)
- `85ee8cf` (UtilMisc revert: comment-only change) — re-ran `UtilMisc` unit tests plus the two classes above: **38 / 0 / 0, BUILD SUCCESS**; CI passed on this head (39/39 checks); Codex re-review: no findings; CodeRabbit: merge risk Minimal, one Trivial JavaDoc finding
- `30c196f` (JavaDoc: comment-only change, no executable lines) — `mvn compile` clean, `UtilMiscUnitTest` **15 / 0 / 0**, encoder lint PASS; CI passed on this head; SonarCloud gate passed with one `S1192` on the pre-existing `"TODAY"` literals
- `cfea87d` (`TODAY_KEYWORD` constant + new `UtilityUnitTest`) — `mvn compile` clean, `UtilityUnitTest` **7 / 0 / 0**, `UtilMiscUnitTest` **15 / 0 / 0**, `scripts/lint/check-bdd-test-naming.sh` PASS, encoder lint PASS; CI re-runs on this head
- `scripts/lint/check-encoder-null-safety.sh` — PASS
- Unicode claims verified against a JVM rather than by inspection: `"falſe".equalsIgnoreCase("false")` → `true`; `"application/x-www-form-urlencoded".length()` → `33`; `"ßa".toUpperCase(ROOT)` → `"SSA"`; `"IŞIK"` → `Işık` under `tr` vs `Işik` under `ROOT`

## Checklist

- [x] My commits are signed off for the [DCO](https://developercertificate.org/) (`git commit -s`)
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/) format, or I've written clear commit messages and will use the format next time
- [x] I have not included any patient data (PHI) in this PR
- [x] I have added tests for new functionality, or this change doesn't need new tests
- [x] I have read the [contributing guide](https://github.com/carlos-emr/carlos/blob/develop/CONTRIBUTING.md)
- [x] I selected the target branch according to the [release process](https://github.com/carlos-emr/carlos/blob/develop/docs/release-process.md)
- [x] I preserved the target branch version/SCM metadata, or this is an explicit release-preparation PR — `pom.xml` remains `2026.08.0-alpha13-SNAPSHOT`, SCM tag `HEAD`
- [x] I did not modify a Flyway migration that exists in a published release tag

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UEqxniWex6MNY6z2mxvYrb

### CODE REVIEW

File: src/main/java/io/github/carlos_emr/carlos/utility/DemographicContactCreator.java [hunk 1]
```diff
@@ -350,10 +350,17 @@ private static ProfessionalSpecialist convertProfessionalContactAsProfessionalSp
...
+        @Override
         public int compare(Contact contact1, Contact contact2) {
-            String lastname1 = contact1.getLastName().toUpperCase();
-            String lastname2 = contact2.getLastName().toUpperCase();
-            return lastname1.compareTo(lastname2);
+            String lastname1 = contact1.getLastName();
+            String lastname2 = contact2.getLastName();
+            if (lastname1 == null) return (lastname2 == null) ? 0 : -1;
+            if (lastname2 == null) return 1;
+            return lastname1.compareToIgnoreCase(lastname2);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/utility/DemographicContactCreator.java]: Confirmed and fixed in 1bac8c2 — this was a real, user-visible regression, and the old ordering was also the *correct* German collation.

Verified on a JVM rather than by reasoning:

```
"ßa".toUpperCase(Locale.ROOT)                       → "SSA"
"ßa".toUpperCase(ROOT).compareTo("TA")              → -1   (ßa before Ta — old behaviour)
"ßa".compareToIgnoreCase("Ta")                      → 107  (ßa after Ta — what this PR had introduced)
Character.toUpperCase('ß')                          → 'ß'  (no single-char mapping, so per-char folding can't reach "SS")
```

The comparator now uses `toUpperCase(Locale.ROOT).compareTo(...)`: locale-independent (the actual goal of this change — `"i".toUpperCase(ROOT)` is `"I"` where a Turkish default locale gives `"İ"`), keeps the new null handling, and preserves the previous folding exactly. A comment at the site records why `compareToIgnoreCase` must not be reintroduced.

Also added `DemographicContactCreatorUnitTest` pinning the contract — sharp s folds to SS, comparison ignores case, nulls sort first — so this can't quietly regress again.

For the record, the `compareToIgnoreCase` form came from an earlier review suggestion on #2425 and is still present on that PR's `develop` branch; it reaches `develop` corrected via the forward-merge of this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@codex review

Re-review requested for the current head `1bac8c2`: the previous Codex review covered `c7fd5d4` only, and `1bac8c2` is the commit that addresses its collation finding (plus `DemographicContactCreatorUnitTest`).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@coderabbitai review

Manual trigger — auto-review skips non-default base branches, and this PR targets `release/2026.08`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/main/java/io/github/carlos_emr/carlos/util/UtilMisc.java [hunk 2]
```diff
@@ -156,7 +157,7 @@ public static String toUpperLowerCase(String S) {
         if (S == null) {
             return S;
         }
-        S = S.trim().toLowerCase();
+        S = S.trim().toLowerCase(Locale.ROOT);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/util/UtilMisc.java]: Confirmed and reverted in 85ee8cf — you're right that this site is display text, and the PR's stated premise ("none of these sites are user-visible display text") was false for it.

Verified both halves before acting:

**Callers.** Every use of `UtilMisc.toUpperLowerCase` formats a patient-facing name: `demographic.getLastName()` / `getFirstName()` / `getAlias()` in `appointmentprovideradminday.jsp`, and the event name in `appointmentgrouprecords.jsp`. Nothing calls it with a protocol token or identifier.

**Locale behaviour, on a JVM** (mirroring the method's lowercase-then-capitalise logic):

```
input                              IŞIK   U+0049 U+015E U+0049 U+004B
tr default (original code)         Işık   U+0049 U+015F U+0131 U+006B   ← dotless ı, correct Turkish
Locale.ROOT (what 1bac8c2 had)     Işik   U+0049 U+015F U+0069 U+006B   ← wrong for Turkish
en-CA default (original code)      Işik   identical to ROOT
```

So on a Canadian deployment the change was invisible, but it was still a real behaviour change for display text — and in the wrong direction for a Turkish-locale JVM.

The site is back to the original default-locale `toLowerCase()`: zero behaviour change on the release line, with a comment recording why it must not be "fixed" again. Picking an explicit display locale is a design decision that doesn't belong in a supported-release fix. The method already carried its `IMPROPER_UNICODE` suppression, so the scanner picture is unchanged.

The other eight files keep `Locale.ROOT`; those really are MIME types, JDBC type names, protocol tokens and identifiers.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@codex review

Re-review requested for the final head `85ee8cf`: the previous Codex review covered `1bac8c2`, and `85ee8cf` is the commit that addresses its `UtilMisc` display-text finding (comment-only revert of that file to the original default-locale call).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
@coderabbitai review

Retry after the earlier rate limit — single manual review of head `85ee8cf` against `release/2026.08` (auto-review is disabled for non-default base branches).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
Re CodeRabbit's one finding on `85ee8cf` (Trivial, outside diff: "Add JavaDoc for the changed public utility methods") — addressed in 30c196f, a **comment-only** commit.

Two corrections to the finding as written, verified against the tree before acting:
- Only `toUpperLowerCase` lives in `UtilMisc.java`. `GetSysDate`, `GetSysDateMin` and `GetSysDateMax` are in `Utility.java`. Both files are already in this PR's diff, so documenting all four stays in scope.
- While writing the contract for `GetSysDateMin`/`GetSysDateMax` I found their fixed-offset parse uses `substring(3, 2)` (begin > end), so every non-blank, non-`TODAY` input throws. There are **no in-tree callers** of either method, so nothing reachable is affected. It is noted inline and deliberately left as-is: a release-line locale fix is not the place to change legacy parsing.

What 30c196f adds: JavaDoc on all four methods covering null/blank handling, the case-insensitive `"TODAY"` sentinel, the parsed shape, the differing failure modes (`GetSysDate` returns `null`; the Min/Max variants throw), and the intentional default-locale casing of `toUpperLowerCase`. It also corrects one sentence in the inline comment I had added to `toUpperLowerCase` in `85ee8cf`: there is no `IMPROPER_UNICODE` suppression on that method (the only one in the file is on `rhtmlEscape`), and none is needed because FindSecBugs only reports a case fold whose result feeds an equality comparison, which this method never does. SpotBugs was and remains green.

Not chasing the 80% docstring-coverage pre-merge threshold beyond this: the remaining undocumented functions in the touched-function set are legacy methods this PR does not otherwise change, and widening a supported-release fix with a JavaDoc sweep is the wrong trade.

Validation on 30c196f: `mvn compile` clean, `UtilMiscUnitTest` 15 / 0 / 0, `scripts/lint/check-encoder-null-safety.sh` PASS. CI re-runs on the new head.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/main/java/io/github/carlos_emr/carlos/util/UtilMisc.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/util/UtilMisc.java]: Correction to my reply above, for the record: I wrote that "the method already carried its `IMPROPER_UNICODE` suppression". That was wrong — `toUpperLowerCase` has no `@SuppressFBWarnings` (the only one in `UtilMisc` is on `rhtmlEscape`), and the inline comment added in 85ee8cf repeated the same claim. The scanner conclusion still holds for a different reason: FindSecBugs only reports a case fold whose result feeds an equality comparison, and this method never compares, so SpotBugs is green without a suppression. The inline comment is corrected and the method documented in 30c196f. Thread stays resolved; the code change it asked for is unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yingbull** [comment]
SonarCloud on 30c196f: Quality Gate passed, but it reported **1 new issue** and a duplication uptick where 85ee8cf had 0 / 1.9%. Both looked at before acting:

- **`java:S1192` in `Utility.java` — `"TODAY"` literal ×3.** Real, and the same finding this PR already fixed in `FrmRecordHelp` for `"TINYINT"`. Fixed in cfea87d: one `TODAY_KEYWORD` constant, three call sites unchanged otherwise (`equalsIgnoreCase` against the same value). Sonar had not reported it on the earlier heads even though the three literals were already there; the issue-tracking evidently lost the match when the JavaDoc shifted those lines and re-raised it as "new". Either way it is legitimately in scope, so it is fixed rather than argued.
- **Duplication 1.9% → 6.8%, all in `UtilMisc.java`.** Sonar counts the changed lines inside `toUpperLowerCase` as duplicated because that whole method is a long-standing twin of `io.github.carlos_emr.Misc.toUpperLowerCase` (the JSPs call both). No code in that block changed in this PR — only comments — so nothing new was duplicated; the metric moved because more changed lines now fall inside the pre-existing duplicate block. Deduplicating two legacy utility classes is not a release-line locale fix. Left as-is.

cfea87d also adds `UtilityUnitTest` (7 tests), pinning the contract documented in 30c196f: `"TODAY"` matched in any case, null/blank → the far-past/far-future sentinels, `d/m/yyyy` parses, `GetSysDate` returns `null` for garbage while the Min/Max variants throw `Invalid Date`. Validated locally: `mvn compile` clean, `UtilityUnitTest` 7 / 0 / 0, `UtilMiscUnitTest` 15 / 0 / 0, BDD-naming lint and encoder lint both PASS. CI re-runs on cfea87d.

---
_Generated by [Claude Code](https://claude.ai/code)_
